File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void check(vector<int>&nums , vector<vector<int>>&v , vector<int>temp , int target , int idx ){
           if(target == 0){
            v.push_back(temp) ; 
            return ; 
        }
        if(target < 0 || idx >= nums.size()){
            return ; 
        } 
     
       
        // temp.push_back(nums[idx]) ; 
        // check(nums , v , temp , target-nums[idx] , idx) ;
        // temp.pop_back() ; 
        // check(nums , v , temp , target , idx+1) ;  
        for(int i = idx ; i<nums.size() ; i++){
         if(i>idx && nums[i] == nums[i-1]) continue ;
        temp.push_back(nums[i]) ; 
        check(nums , v , temp , target-nums[i] , i+1) ;
        temp.pop_back() ; 
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& nums, int target) {
        sort(nums.begin() , nums.end()) ; 
        vector<int>temp ; 
        vector<vector<int>>v; 

        check(nums , v , temp , target , 0) ; 
        return v ; 
    }
};

File: 0040-combination-sum-ii/0040-combination-sum-ii.java
class Solution {
      void check(int[] nums , int idx , List<Integer>temp , int target , List<List<Integer>>v){
         if(target == 0){
             v.add(new ArrayList<>(temp)) ; 
             return ;
        }
        if(idx>= nums.length || target < 0) {
            return ; 
        }
       
        
        // temp.add(nums[idx]) ; 
        // check(nums , idx , temp , target - nums[idx], v) ; 
        // temp.remove(temp.size()-1) ; 
        // check(nums , idx+1 ,temp , target , v) ; 

        for(int i = idx ;i<nums.length ; i++){
            if(i>idx && nums[i] == nums[i-1]) continue ;
            temp.add(nums[i]) ; 
            check(nums , i+1 , temp , target-nums[i] , v) ; 
            temp.remove(temp.size()-1) ; 
        }
       
    }
    public List<List<Integer>> combinationSum2(int[] nums, int target) {
         List<List<Integer>>v = new ArrayList<>() ; 
        List<Integer>temp = new ArrayList<>() ; 
        Arrays.sort(nums) ; 
        check(nums , 0 , temp , target  , v) ; 

        return v ; 
    }
}